app/models/transformation_mapping_item.rb

```diff
@@ -3,14 +3,20 @@ class TransformationMappingItem < ApplicationRecord
   belongs_to :source,      :polymorphic => true
   belongs_to :destination, :polymorphic => true
 
-  validates :source_id, :uniqueness => {:scope => [:transformation_mapping_id, :source_type]}
+  validates :source_id, :uniqueness => {:scope => [:transformation_mapping_id, :source_type, :destination_type] }
 
   validate :source_cluster,      :if => -> { source.kind_of?(EmsCluster) }
   validate :destination_cluster, :if => -> { destination.kind_of?(EmsCluster) || destination.kind_of?(CloudTenant) }
 
+  validate :source_datastore,    :if => -> { source.kind_of?(Storage) }
```

Comment:
The more I read through this, the more I think this would benefit from subclassing for specific mapping types (storage, cluster, etc.)

Reply:
(Probably not for this PR though)

Reply:
:+1:

Reply:
sure.